examples/fit-model-ceres.cpp

```diff
@@ -83,406 +262,159 @@ std::ostream& operator<<(std::ostream& out, const std::vector<T>& v)
-                                             camera_translation_and_intrinsics[1],
-                                             camera_translation_and_intrinsics[2]));
-        } else
-        {
-            return glm::translate(
-                glm::dvec3(camera_translation_and_intrinsics[0], camera_translation_and_intrinsics[1], 0.0));
-        }
-    };
-    auto get_projection_matrix = [](auto&& camera_translation_and_intrinsics, auto&& aspect,
-                                    auto&& use_perspective) {
-        if (use_perspective)
-        {
-            const auto& focal = camera_translation_and_intrinsics[3];
-            return glm::perspective(focal, aspect, 0.1, 1000.0);
-        } else
-        {
-            const auto& frustum_scale = camera_translation_and_intrinsics[2];
-            return glm::ortho(-1.0 * aspect * frustum_scale, 1.0 * aspect * frustum_scale,
-                              -1.0 * frustum_scale, 1.0 * frustum_scale);
-        }
-    };
-    auto t_mtx = get_translation_matrix(camera_translation_and_intrinsics, use_perspective);
-    auto projection_mtx = get_projection_matrix(camera_translation_and_intrinsics, aspect, use_perspective);
-    const glm::dvec4 viewport(0, image.rows, image.cols, -image.rows); // OpenCV convention
-
-    auto mean_mesh = morphable_model.get_mean();
-    for (auto idx : vertex_indices)
-    {
-        glm::dvec3 point_3d(mean_mesh.vertices[idx][0], mean_mesh.vertices[idx][1],
-                            mean_mesh.vertices[idx][2]); // The 3D model point
-        glm::dvec3 projected_point = glm::project(point_3d, t_mtx * rot_mtx, projection_mtx, viewport);
-        cv::circle(outimg, cv::Point2f(projected_point.x, projected_point.y), 3, {0.0f, 0.0f, 255.0f}); // red
-    }
-    for (const auto& lm : image_points)
-    {
-        cv::circle(outimg, cv::Point2f(lm(0), lm(1)), 3,
-                   {0.0f, 255.0f, 255.0f}); // yellow: subset of the detected LMs that we use
-    }
-    auto euler_angles = glm::eulerAngles(estimated_rotation); // returns [P, Y, R]
-    fitting_log << "Pose fit with mean shape:\tYaw " << glm::degrees(euler_angles[1]) << ", Pitch "
-                << glm::degrees(euler_angles[0]) << ", Roll " << glm::degrees(euler_angles[2])
-                << "; t & f: " << camera_translation_and_intrinsics << '\n';
-    fitting_log << "Ceres took: "
-                << std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() << "ms.\n";
+
+    Mat outimg = image.clone();
+    ceres_example::draw_mesh_vertices(outimg, morphable_model.get_mean(), indexed_landmarks,
+                                      camera.calculate_translation_matrix() * camera.calculate_rotation_matrix(),
+                                      camera.calculate_projection_matrix(), camera.get_viewport());
+    ceres_example::draw_landmarks(outimg, indexed_landmarks);
+
+    auto camera_euler_rotation = camera.get_euler_rotation();
```

Comment:
I think this can probably be const?

Reply:
You are right

Reply:
I didn't make it `const` because I use this variable second time (here: https://github.com/patrikhuber/eos/pull/248/files/8c6137ce4ae423e6b9dd231e61e8d9536a6815e7#diff-1a39125f7bf324839e08188060f75b96R402). Do you think it is better to make another one?